### PR TITLE
feat: add onlinefix scraper

### PR DESCRIPTION
**Pull request by pmenta**
Adding the onlinefix as a new repacker source.

- There is need to set new env values with onlinefix credentials
- Only crawls the homepage of onlinefix

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -93,11 +96,14 @@
...
     "uuid": "^9.0.1",
     "vite-plugin-svgr": "^4.2.0",
     "vite-tsconfig-paths": "^4.3.2",
-    "winston": "^3.12.0"
+    "windows-1251": "^3.0.4",
+    "winston": "^3.12.0",
+    "zod": "^3.22.4"
   }
-}
+}
```
Comment (hydralauncher) [inline package.json:109]: No newline at the end

File: src/main/events/helpers/search-games.ts [hunk 2]
```diff
@@ -13,11 +13,15 @@ const repacksIndex = new Index();
...
-
-  repacksIndex.add(i, formatName(formatter(repack.title)));
+  try {
+    const repack = repacks[i];
+    const formatter =
+      repackerFormatter[repack.repacker as keyof typeof repackerFormatter];
+
+    repacksIndex.add(i, formatName(formatter(repack.title)));
+  } catch (err) {
+    console.log(err);
```
Comment (hydralauncher) [inline src/main/events/helpers/search-games.ts:23]: Please remove the `console.log`

File: src/main/helpers/formatters.ts [hunk 3]
```diff
@@ -47,6 +47,7 @@ export const xatabFormatter = (title: string) =>
     .replace(/(v\.?([0-9]| )+)+([0-9]|\.|-|_|\/|[a-zA-Z]| )+/, "");
 
 export const tinyRepacksFormatter = (title: string) => title;
+export const onlinefixFormatter = (title: string) => title;
```
Comment (hydralauncher) [inline src/main/helpers/formatters.ts:50]: Please add a proper formatter for the repacker if necessary. Any repacker that post their games with anything that goes beyond the game title should have its own RegExp that will force the title to be formatted like a regular game title. You can check FitGirl's one as an example.

File: package.json [hunk 4]
```diff
@@ -76,15 +77,17 @@
...
+    "date-fns": "^3.6.0",
     "electron-squirrel-startup": "^1.0.0",
     "flexsearch": "^0.7.43",
+    "got-scraping": "^4.0.5",
     "i18next": "^23.10.1",
     "i18next-browser-languagedetector": "^7.2.0",
     "jsdom": "^24.0.0",
     "lodash": "^4.17.21",
-    "parse-torrent": "9.1.5",
+    "parse-torrent": "^9.1.5",
```
Comment (hydralauncher) [inline package.json:88]: Please keep the `pase-torrent` version locked due to the low compatibility of newer versions on `win32`

File: package.json [hunk 5]
```diff
@@ -93,11 +96,14 @@
...
+    "tough-cookie": "^4.1.3",
     "typeorm": "^0.3.20",
     "update-electron-app": "^3.0.0",
     "uuid": "^9.0.1",
     "vite-plugin-svgr": "^4.2.0",
     "vite-tsconfig-paths": "^4.3.2",
-    "winston": "^3.12.0"
+    "windows-1251": "^3.0.4",
+    "winston": "^3.12.0",
+    "zod": "^3.22.4"
```
Comment (hydralauncher) [inline package.json:107]: I see no need for the addition of `zod` in the context of this project

File: src/main/services/repack-tracker/online-fix.ts [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+          .get(gameLink, {
+            encoding: "binary",
+          })
+          .text();
+
+        const gameDom = new JSDOM(gamePage);
+        const gameDocument = gameDom.window.document;
+
+        const uploadDateText = gameDocument.querySelector(
+          "#dle-content > div > article > div.full-story-header.wide-block.clr > div.full-story-top-panel.clr > div.date.left > time"
```
Comment (hydralauncher) [inline src/main/services/repack-tracker/online-fix.ts:103]: This selector seems way to specific. Can we make it more generic?

File: src/main/services/repack-tracker/online-fix.ts [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        ).textContent;
+
+        let decodedDateText = decode(uploadDateText);
+
+        // "Вчера" significa ontem.
```
Comment (hydralauncher) [inline src/main/services/repack-tracker/online-fix.ts:108]: English comments only

File: src/main/services/repack-tracker/online-fix.ts [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+          .text();
+
+        const torrentDom = new JSDOM(torrentPage);
+        const torrentDocument = torrentDom.window.document;
+
+        const torrentLink = torrentDocument
+          .querySelector("a:nth-child(2)")
+          ?.getAttribute("href");
+
+        const torrentFile = Buffer.from(
```
Comment (hydralauncher) [inline src/main/services/repack-tracker/online-fix.ts:152]: Please make the already existing function to download files as Buffers into a helper and use it here:
https://github.com/hydralauncher/hydra/blob/main/src/main/services/repack-tracker/xatab.ts#L10

File: src/main/services/repack-tracker/online-fix.ts [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+        const torrentPage = await http
+          .get(torrentPrePage, {
+            encoding: "binary",
+            headers: {
+              Referer: gameLink,
+            },
+          })
+          .text();
+
+        const torrentDom = new JSDOM(torrentPage);
```
Comment (hydralauncher) [inline src/main/services/repack-tracker/online-fix.ts:145]: Please separate every new instantiation of `JSDOM` in a new function so you can provide better readability for the interaction with the `document` object

File: src/main/services/repack-tracker/online-fix.ts [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+            .buffer()
+        );
+
+        const torrent = parseTorrent(torrentFile);
+        const magnetLink = toMagnetURI({
+          infoHash: torrent.infoHash,
+        });
+
+        repacks.push({
+          fileSize: "NA",
```
Comment (hydralauncher) [inline src/main/services/repack-tracker/online-fix.ts:166]: Is there no way to know the `fileSize` for online-fix?

File: package-lock.json
Comment (hydralauncher) [inline package-lock.json:1]: Please don't add multiple locks. Keep it with `yarn.lock`

**Review by hydralauncher — CHANGES_REQUESTED** [review]

File: src/main/services/repack-tracker/online-fix.ts (hunk 8 above)
Comment (pmenta) [reply src/main/services/repack-tracker/online-fix.ts:152]: The existing function dont has support to cookieJar and header-generation that is needed to download the torrent file.

File: src/main/services/repack-tracker/online-fix.ts (hunk 10 above)
Comment (pmenta) [reply src/main/services/repack-tracker/online-fix.ts:166]: No.

File: package.json [hunk 11]
```diff
@@ -93,11 +96,14 @@
...
     "sqlite3": "^5.1.7",
     "systeminformation": "^5.22.3",
+    "tough-cookie": "^4.1.3",
     "typeorm": "^0.3.20",
     "update-electron-app": "^3.0.0",
     "uuid": "^9.0.1",
     "vite-plugin-svgr": "^4.2.0",
     "vite-tsconfig-paths": "^4.3.2",
-    "winston": "^3.12.0"
+    "windows-1251": "^3.0.4",
```
[the remark this answers is not in the document]
Comment (pmenta) [reply package.json:105]: To decode the cyrlic characters returned in the onlinefix html.

[the remark this answers is not in the document]
Comment (hydralauncher) [reply package.json:105]: It's always better to decode/format repack titles on runtime. The reason is that this is more efficient when discovering which repacks have been already added to the table.

**Pull request closed without merging**